casper/src/main/resources/RevVault.rho

```diff
@@ -32,60 +32,64 @@ in {
       for (mint <- mintCh) {
 
         //FIXME the result of findOrCreateGenesisVault is not protected, nor is the call to it. Both should be.
-        contract RevVault(@"findOrCreateGenesisVault", @ownerPubKeyBytes, @initialAmount, ret) = {
+        contract RevVault(@"findOrCreateGenesisVault", @ownerRevAddress, @initialAmount, ret) = {
           new findOrCreateGenesisVault in {
-            _getOrCreate!(ownerPubKeyBytes, *findOrCreateGenesisVault, *ret) |
+            _getOrCreate!(ownerRevAddress, *findOrCreateGenesisVault, *ret) |
             for (retCh <- findOrCreateGenesisVault) {
 
-              _makeVault!(ownerPubKeyBytes, initialAmount, *retCh)
+              _makeVault!(ownerRevAddress, initialAmount, *retCh)
             }
           }
         } |
 
-        contract RevVault(@"findOrCreate", @ownerPubKeyBytes, ret) = {
+        contract RevVault(@"findOrCreate", @ownerRevAddress, ret) = {
 
           new createLockedVault in {
-            _getOrCreate!(ownerPubKeyBytes, *createLockedVault, *ret) |
+            _getOrCreate!(ownerRevAddress, *createLockedVault, *ret) |
             for (retCh <- createLockedVault) {
 
               new vaultCh, lock in {
-                _makeVault!(ownerPubKeyBytes, 0, *vaultCh) |
+                _makeVault!(ownerRevAddress, 0, *vaultCh) |
                 @Either!("map <-", *vaultCh, *lock, *retCh) |
                 for (@vault, retCh <- lock) {
-                  @Lockbox!("lock", ownerPubKeyBytes, vault, *retCh)
+                  @Lockbox!("lock", ownerRevAddress, vault, *retCh)
                 }
               }
             }
           }
         } |
 
-        contract _getOrCreate(@key, constructor, ret) = {
-          for (@vaultMap <- vaultMapStore) {
-            match vaultMap.get(key) {
-              Nil => {
-                new resCh in {
-                  constructor!(*resCh) |
-                  for (@eitherVault <- resCh) {
-                    match eitherVault {
-                      ("Right", vault) => vaultMapStore!(vaultMap.set(key, vault))
-                      ("Left",  _    ) => vaultMapStore!(vaultMap)
-                    } |
-                    ret!(eitherVault)
+        contract _getOrCreate(@revAddress, constructor, retCh) = {
+          new revAddressValidCh, revAddressValidEitherCh, getOrCreate in {
+            RevAddress!("validate", revAddress, *revAddressValidCh) |
+            @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
+            @Either!("flatMap <-", *revAddressValidEitherCh, *getOrCreate, *retCh) |
```

Comment:
You don't really need `retCh` here, do you?

Reply:
WDYM? I need to signal to the caller that flatMap finished. It might be that it's a left, in which case the 'lambda' won't even be called and won't have a chance to signal a return. Not to mention I should never rely on that :)

Reply:
Or the next line should be: `for (_, ret <- retCh; @vaultMap <- vaultMapStore) {` ?

Reply:
no, it shouldn't :) `getOrCreate` is the handle under which a "lambda" to be called by `flatMap` is going to listen. As per the "method call convention" we seem to have developed, the labmda takes as many parameters it needs, plus a return channel. But this is a different return channel than `flatMap` will return to. In fact, each "method caller" must spawn its own return channel and pass it to the "called method".

Reply:
Thanks, I see it now.